csrc/scheduler/normalization_utils.cpp

```diff
@@ -813,6 +813,47 @@ void checkReductionTvForScheduling(Fusion* fusion, TensorView* ref_red_tv) {
       "Tried to schedule a fusion with no tensor inputs, currently not supported.");
 }
 
+// Returns true if the gains of reducing buffer size is larger than the pains of
+// recalculations. We don't know the real answer until we run it.
+bool projectBufferToInputs(
+    Fusion* fusion,
+    const scheduler_utils::PersistentBufferInfo& persistent_buffer_info,
+    const scheduler_utils::PersistentBufferSizeReturn&
+        persistent_buffer_size_info) {
+  // don't project if can't reduce buffer size
+  if (persistent_buffer_size_info.projected_persistent_buffer_size >=
+      persistent_buffer_size_info.persistent_buffer_size) {
+    return false;
+  }
+
+  // check ops between persistent buffer and inputs.
+  // TODO: check more ops, e.g. RNGOp
+  bool has_exp_op = false;
+  const auto& projectable_buffers =
+      persistent_buffer_info.projectable_persistent_buffers;
+  auto all_inputs = ir_utils::inputTvsOf(projectable_buffers);
+  const auto all_exprs = StmtSort::getExprsBetween(
+      {all_inputs.begin(), all_inputs.end()},
+      {projectable_buffers.begin(), projectable_buffers.end()});
+  for (auto expr : all_exprs) {
+    if (expr->isA<UnaryOp>() &&
+        expr->as<UnaryOp>()->getUnaryOpType() == UnaryOpType::Exp) {
+      has_exp_op = true;
+    }
+  }
+  if (!has_exp_op) {
+    return true;
+  }
+
+  // For cases with expensive ops, don't project if buffer size is 'small',
+  if (persistent_buffer_size_info.persistent_buffer_size <=
+      scheduler_utils::small_buffer_size_threshold) {
```

Comment:
Can we avoid using the magic value? The underlying issue is that if the non-projected persistent buffer is small enough, i.e., not affecting the occupancy, projecting back to the inputs isn't buying us anything. Can we infer when it becomes small enough?

Reply:
Also, can we just do this even without expensive operations? The perf difference would be small but it shouldn't result in lower performances, right?

Reply:
> Can we avoid using the magic value? The underlying issue is that if the non-projected persistent buffer is small enough, i.e., not affecting the occupancy, projecting back to the inputs isn't buying us anything. Can we infer when it becomes small enough?

(1) Probabally no. To calculate occupancy, we need to estimate register usage which is based on magic number (24 or 40 registers as overhead).
(2) Calculate occupancy is also non-trivial, we need to run the heuristics with project_to_input = True / False, then compare the change of occupancy.

Reply:
> Also, can we just do this even without expensive operations? The perf difference would be small but it shouldn't result in lower performances, right?

we can do that. You are right, the perf difference would be small. Maybe we just don't need to change it if the difference is small. we can leave it for a following work and this PR introduces this function as a placehoder and only improve softmax without causing regressions on other cases.

Reply:
> (1) Probabally no. To calculate occupancy, we need to estimate register usage which is based on magic number (24 or 40 registers as overhead).

Why does this matter? That magic number already exists. Why not avoid adding another if not strictly required?